### PR TITLE
Elementalist

### PR DESCRIPTION
**Pull request by minin43**
- [x] Dynamic HTML-based tutorial
- [x] `CreateConVar` result caching to remove `GetConvar` usage in equipment item creation
- [x] Verify elementalist upgrade ("+") convar functionality
- [x] Resolve luacheck errors

### CODE REVIEW

File: lua/customroles/elementalist/cl_elementalist.lua [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        ["$pp_colour_colour"] = 1
+    }
+end
+
+hook.Add("RenderScreenspaceEffects", "ElementalistScreenDimming", function()
+    DrawColorModify(DarkOverlay)
+end)
+
+net.Receive("BeginIceScreen", function(len)
+    local frozen = net.ReadBool()
```
Comment (Malivil) [inline lua/customroles/elementalist/cl_elementalist.lua:77]: Is this supposed to do something?

File: lua/customroles/elementalist/sv_elementalist.lua [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+    if not IsValidPlayerEnt(ent) or not IsValidPlayerEnt(att) or not dmginfo:IsBulletDamage() then -- For some reason, crowbar attacks return true on this
+        return
+    end
+
+    if not att:IsRole(ROLE_ELEMENTALIST) then
+        return
+    end
+
+    -- Att is a valid elementalist damaging a valid player
+    local attId = att:SteamID64()
```
Comment (Malivil) [inline lua/customroles/elementalist/sv_elementalist.lua]: This is not used

File: lua/customroles/elementalist/sv_elementalist.lua [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+                explosion:SetOwner(att)
+                explosion:Spawn()
+                explosion:SetKeyValue("iMagnitude", ent:Health() * 2)
+                explosion:Fire("Explode", 0, 0)
+                util.BlastDamage(explosion, att, ent:GetPos(), ent:Health() * 4, ent:Health()) 
+            end
+        else
+            --Base functionality
+            IgnitedPlayers[vicId] = true
+            local timerID = vicId .. "_IsBurning"
```
Comment (Malivil) [inline lua/customroles/elementalist/sv_elementalist.lua]: Casing is different from usage below
```suggestion
            local timerId = vicId .. "_IsBurning"
```

File: lua/customroles/elementalist/sv_elementalist.lua [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+            --Upgrade functionality
+            BlindedPlayers[vicId] = 100
+        else
+            --Base functionality
+            BlindedPlayers[vicId] = BlindedPlayers[vicId] or 0
+            BlindedPlayers[vicId] = math.Clamp(BlindedPlayers[vicId] + damage, 0, 50)            
+        end
+
+        net.Start("BeginDimScreen")
+            net.WriteUInt(BlindedPlayers[vidId], 6)
```
Comment (Malivil) [inline lua/customroles/elementalist/sv_elementalist.lua]:
```suggestion
            net.WriteUInt(BlindedPlayers[vicId], 6)
```

File: lua/customroles/elementalist/cl_elementalist.lua [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    else
+        StartDarkOverlay(amount)
+    end
+end)
+
+net.Receive("EndDimScreen", function(len)
+    EndDarkOverlay()
+end)
+
+hook.Add("TTTTutorialRoleText", "SummonerTutorialRoleText", function(role)
```
Comment (Malivil) [inline lua/customroles/elementalist/cl_elementalist.lua]: Tutorials are usually HTML-based descriptions with multiple sentences and color-based highlighting.
They also generally talk about the different features of the role (in this case, the different powers that can be bought) and dynamically change based on the convar values (if there are any).

For an example, see here: https://github.com/Custom-Roles-for-TTT/TTT-Custom-Roles/blob/beta/gamemodes/terrortown/gamemode/roles/arsonist/cl_arsonist.lua#L279

File: lua/customroles/elementalist/shared.lua [hunk 6]
```diff
@@ -0,0 +1,292 @@
+--// Logan Christianson
+-- Allowing of individual effect upgrades
+CreateConVar("ttt_elementalist_allow_effect_upgrades", "1", FCVAR_REPLICATED, "Controls whether elemental effect \"upgrades\" should be available for purchase", 0, 1)
```
Comment (Malivil) [inline lua/customroles/elementalist/shared.lua]: Are all of these used on the client as well? Only convars that are used on the client (even just for the tutorial) are replicated. Any that are only used on the server should only be created on the server and no have the replicated flag.

File: lua/customroles/elementalist/shared.lua [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+CreateConVar("ttt_elementalist_allow_windburn_upgrades", "1", FCVAR_REPLICATED, "Controls whether the Windburn upgrade(s) should be available in the elementalist shop", 0, 1)
+CreateConVar("ttt_elementalist_allow_discharge_upgrades", "1", FCVAR_REPLICATED, "Controls whether the Discharge upgrade(s) should be available in the elementalist shop", 0, 1)
+CreateConVar("ttt_elementalist_allow_midnight_upgrades", "1", FCVAR_REPLICATED, "Controls whether the Midnight upgrade(s) should be available in the elementalist shop", 0, 1)
+CreateConVar("ttt_elementalist_allow_lifesteal_upgrades", "1", FCVAR_REPLICATED, "Controls whether the Lifesteal upgrade(s) should be available in the elementalist shop", 0, 1)
+
+-- Durations & associated chances
+CreateConVar("ttt_elementalist_frostbite_effect_duration", "3", FCVAR_REPLICATED, "How long the Frostbite slow & freeze effect lasts. Value must be greater than 0 and less than 6", 1, 5)
+CreateConVar("ttt_elementalist_frostbite+_freeze_chance", "5", FCVAR_REPLICATED, "The percent chance shooting a victim which has been slowed by Frostbite will instead freeze them. Value must be greater than 0 and less than 101", 1, 100)
```
Comment (Malivil) [inline lua/customroles/elementalist/shared.lua]: Have you tested convars with symbols in them before?
These may need to be changed to "frostbiteplus" etc.

File: lua/customroles/elementalist/shared.lua [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+-- end
+
+RegisterRole(ROLE)
+
+if SERVER then
+    AddCSLuaFile()
+end
+
+hook.Add("Initialize", "Elementalist Shop Items Init", function()
+    local getConVar = GetConVar
```
Comment (Malivil) [inline lua/customroles/elementalist/shared.lua]: Why?
If you're worried about performance, the best thing to do would be to save the convar above where it's created and then use that variable here so you don't need to call GetConVar at all

File: lua/customroles/elementalist/sv_elementalist.lua [hunk 9]
```diff
@@ -0,0 +1,284 @@
+--// Logan Christianson
+util.AddNetworkString("BeginIceScreen")
+util.AddNetworkString("EndIceScreen")
+util.AddNetworkString("BeginDimScreen")
+util.AddNetworkString("EndDimScreen")
+
+local GetConVar = GetConVar
+
+local function IsValidPlayerEnt(ent)
+    return ent and IsValid(ent) and ent:IsPlayer() and ent:Alive()
```
Comment (Malivil) [inline lua/customroles/elementalist/sv_elementalist.lua]:
```suggestion
    return IsPlayer(ent) and ent:Alive()
```

File: lua/customroles/elementalist/sv_elementalist.lua [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+end)
+
+hook.Add("EntityTakeDamage", "Elementalist Effects", function(ent, dmginfo)
+    local att = dmginfo:GetAttacker()
+
+    if not IsValidPlayerEnt(ent) or not IsValidPlayerEnt(att) or not dmginfo:IsBulletDamage() then -- For some reason, crowbar attacks return true on this
+        return
+    end
+
+    if not att:IsRole(ROLE_ELEMENTALIST) then
```
Comment (Malivil) [inline lua/customroles/elementalist/sv_elementalist.lua]:
```suggestion
    if not att:IsElementalist() then
```

File: lua/customroles/elementalist/sv_elementalist.lua [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+            edata:SetEntity(ent)
+            edata:SetMagnitude(3)
+            edata:SetScale(2)
+
+            util.Effect("TeslaHitBoxes", edata)
+        end
+
+        local eyeang = ent:EyeAngles()
+        if eyeang then
+            local j = GetConVar("ttt_elementalist_discharge_punch_power"):GetInt() * damage
```
Comment (Malivil) [inline lua/customroles/elementalist/sv_elementalist.lua]: Can we change this to a meaningful variable name?

**Review by Malivil — CHANGES_REQUESTED** [review]
There are some errors in here that makes me think this has not been fully tested

Also, please update the README to include information about your role. See here for last year's for an example: https://github.com/Custom-Roles-for-TTT/TTT-Jingle-Jam-Roles-2022/blob/main/README.md

File: lua/customroles/elementalist/cl_elementalist.lua (hunk 1 above)
Comment (minin43) [reply lua/customroles/elementalist/cl_elementalist.lua:77]: Yea that's a goof. That value is supposed to be passed into the function and determines how intense the overlay effect is, forgot to actually add that. Will resolve when fixed

File: lua/customroles/elementalist/shared.lua (hunk 6 above)
Comment (minin43) [reply lua/customroles/elementalist/shared.lua]: Absolutely none of them are used on the client, I have no idea why I kept that flag for all of them. Will resolve comment when fixed.

File: lua/customroles/elementalist/shared.lua (hunk 8 above)
Comment (minin43) [reply lua/customroles/elementalist/shared.lua]: You know, I never thought that CreateConVar returns the ConVar object itself I can then re-use elsewhere... I suppose I'll update that.

File: lua/customroles/elementalist/sv_elementalist.lua (hunk 2 above)
Comment (minin43) [reply lua/customroles/elementalist/sv_elementalist.lua]: Huh, really, guess it isn't. Will resolve when removed.

File: lua/customroles/elementalist/sv_elementalist.lua [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+
+        if not timer.Exists(vicId .. "_IsSlowed") then
+            timer.Create(vicId .. "_IsSlowed", Timer, 1, EndFrostbite)
+        else
+            timer.Adjust(vicId .. "_IsSlowed", Timer, 1, EndFrostbite)
+        end
+    end
+
+    if att:HasEquipmentItem(EQUIP_ELEMENTALIST_PYROMANCER) then
+        local timerId = vicId .. "_isBurning"
```
Comment (Malivil) [inline lua/customroles/elementalist/sv_elementalist.lua]: Remove this in favor of the one on line 114 with correct capitalization

**Review by Malivil — CHANGES_REQUESTED** [review]

**Comment by Malivil** [comment]
Only remaining static error is not due to your code so it can be ignored. I can bypass the check when we merge.
Other than outstanding comments, we're looking good =)

**Comment by minin43** [comment]
The Windburn+ effect launchers terrorists into the air, ideally it would block attempts to double-jump and land safely, however, there's probably a variety of potential double jump addons a server might have installed. Would it be bad practice to check for and disable the jump for most common addons when the effect occurs? Or just leave it as-is?

**Comment by Malivil** [comment]
> The Windburn+ effect launchers terrorists into the air, ideally it would block attempts to double-jump and land safely, however, there's probably a variety of potential double jump addons a server might have installed. Would it be bad practice to check for and disable the jump for most common addons when the effect occurs? Or just leave it as-is?

If you want that as a feature that's how I would do it. For my version, you would do something like:

```lua
if ply.SetJumpLevel and ply.GetMaxJumpLevel then
    ply:SetJumpLevel(ply:GetMaxJumpLevel() + 1)
end
```

That should work for at least these versions:
- https://steamcommunity.com/sharedfiles/filedetails/?id=2501234496
- https://steamcommunity.com/sharedfiles/filedetails/?id=284538302
- https://steamcommunity.com/sharedfiles/filedetails/?id=1962801891

**Comment by minin43** [comment]
Everything else should be green, only thing to do now is test the convars work as expected

File: lua/customroles/elementalist/sv_elementalist.lua [hunk 13]
```diff
@@ -138,6 +138,10 @@ hook.Add("EntityTakeDamage", "Elementalist Effects", function(ent, dmginfo)
         if att:HasEquipmentItem(EQUIP_ELEMENTALIST_WINDBURN_UP) and GetChanceConVarOutcome("ttt_elementalist_windburn+_launch_chance") then
             --Upgrade functionality
             ent:SetVelocity(ent:GetVelocity() + Vector(0, 0, 1000 + (1000 * scale)))
+
+            if ply.SetJumpLevel and ply.GetMaxJumpLevel then
```
Comment (Malivil) [inline lua/customroles/elementalist/sv_elementalist.lua]: Did you test this? I'm wondering if there needs to be a slight delay so the player is actually in the air

File: lua/customroles/elementalist/cl_elementalist.lua [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+    else
+        StartDarkOverlay(amount)
+    end
+end)
+
+net.Receive("EndDimScreen", function(len)
+    EndDarkOverlay()
+end)
+
+hook.Add("TTTTutorialRoleText", "SummonerTutorialRoleText", function(playerRole)
```
Comment (Malivil) [inline lua/customroles/elementalist/cl_elementalist.lua:100]:
```suggestion
hook.Add("TTTTutorialRoleText", "ElementalistTutorialRoleText", function(playerRole)
```

File: lua/customroles/elementalist/cl_elementalist.lua [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+    local function getStyleString(role)
+        local roleColor = ROLE_COLORS[role]
+        return "<span style='color: rgb(" .. roleColor.r .. ", " .. roleColor.g .. ", " .. roleColor.b .. ")'>"
+    end
+
+    if playerRole == ROLE_ELEMENTALIST then
+        local divStart = "<div style='margin-top: 10px;'>"
+        local styleEnd = "</span>"
+
+        local html = "The " .. ROLE_STRINGS[ROLE_ELEMENTALIST] .. " is a member of the " .. getStyleString(ROLE_TRAITOR) .. ROLE_STRINGS_EXT[ROLE_TRAITOR] .. " team" .. styleEnd .. " whose goal is to eliminate all innocents and independents."
```
Comment (Malivil) [inline lua/customroles/elementalist/cl_elementalist.lua]: Don't use role constants for team names. 
Also the _EXT role string adds an article (e.g. a Traitor) which is not correct here

```suggestion
        local html = "The " .. ROLE_STRINGS[ROLE_ELEMENTALIST] .. " is a member of the " .. getStyleString(ROLE_TRAITOR) .. "traitor team" .. styleEnd .. " whose goal is to eliminate all innocents and independents."
```

**Review by Malivil — CHANGES_REQUESTED** [review]

File: lua/customroles/elementalist/cl_elementalist.lua (hunk 15 above)
Comment (minin43) [reply lua/customroles/elementalist/cl_elementalist.lua]: What the difference between this context and the change you request in the Physician PR?

Comment (Malivil) [reply lua/customroles/elementalist/cl_elementalist.lua]: This is talking about the traitor team which cannot be renamed while that was talking about the detective role which can be renamed.

Comment (minin43) [reply lua/customroles/elementalist/cl_elementalist.lua]: Ah, okay, fair enough.

**Comment by Malivil** [comment]
There are luacheck errors to fix as well

**Comment by minin43** [comment]
> There are luacheck errors to fix as well

I don't see them, for me it still says it's waiting

File: lua/customroles/elementalist/sv_elementalist.lua (hunk 13 above)
Comment (minin43) [reply lua/customroles/elementalist/sv_elementalist.lua]: Can't really test without another person, figured there'd be some general playtesting done before releasing the pack and I could test this and other functionality then.

Comment (Malivil) [reply lua/customroles/elementalist/sv_elementalist.lua]: > Can't really test without another person, figured there'd be some general playtesting done before releasing the pack and I could test this and other functionality then.

If you do `bot_mimic 1` they will jump when you jump so you can test that way

**Comment by Malivil** [comment]
> > There are luacheck errors to fix as well
> 
> I don't see them, for me it still says it's waiting

Yea, annoyingly it hides the previous run when you commit again. Here's the link to it: https://github.com/Custom-Roles-for-TTT/TTT-Jingle-Jam-Roles-2023/actions/runs/7448619902/job/20264014601

File: lua/customroles/elementalist/sv_elementalist.lua (hunk 13 above)
Comment (minin43) [reply lua/customroles/elementalist/sv_elementalist.lua]: MIMIC bot? I had no idea such a thing existed.

There's a small handful of other things that will still need to be tested regardless if I can prove the double jump removal works or not. Is it okay to wait or do you adamantly want everything that can be tested/working tested and working before any playtesting?

Comment (Malivil) [reply lua/customroles/elementalist/sv_elementalist.lua]: We don't normally do a formal test of things, it's more like we release it to our friend group and try it out there. 

We'd like to make sure the basic functionality works before then so we're not spending time fixing errors while they just want to play. 

The timeline is very relaxed here so it doesn't need to be done right away. Also I'm probably going to test things out myself just so I'm familiar with the roles anyway so it's not a big deal if you can't test right now.

Comment (minin43) [reply lua/customroles/elementalist/sv_elementalist.lua]: Yea I'm not anticipating anything BREAKS, only that I haven't seen how certain effects will work when playing, if certain inputs will actually register, how screen overlays look, etc. If you're content waiting, feel free to resolve the conversation. I went ahead and added a tick wait anyway.

**Review by Malivil — APPROVED** [review]
LGTM, assuming convars with +s work

**Comment by minin43** [comment]
Added final change to only call effect rendering when client should - tested convar names, mine with the `+` work as expected.

**Review by Malivil — APPROVED** [review]
LGTM

Only luacheck error is it looking for the gamemodes folder which we can ignore

File: materials/vgui/ttt/roles/elm/elementalist.png
Comment (NoxxFlame) [inline materials/vgui/ttt/roles/elm/elementalist.png:1]: I'm being super nitpicky here but I wonder if this icon would be better without the 3 element symbols. It looks really cool when its large but its never going to be anywhere big enough for them to be recognizable.

File: lua/customroles/elementalist/cl_elementalist.lua [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+        local html = "The " .. ROLE_STRINGS[ROLE_ELEMENTALIST] .. " is a member of the " .. getStyleString(ROLE_TRAITOR) .. "traitor team" .. styleEnd .. " whose goal is to eliminate all innocents and independents."
+
+        html = html .. divStart .. "They have access to " .. getStyleString(ROLE_ELEMENTALIST) .. "special powerups" .. styleEnd .. " in the " .. getStyleString(ROLE_TRAITOR) .. "traitor shop" .. styleEnd .. " which do a variety of things when they shoot other terrorists.</div>"
+
+        html = html .. divStart .. "The powerups available are:<ul style='margin-bottom: 0px; padding-bottom: 0px;'>"
+
+        local ROLE = ROLE_DATA_EXTERNAL[ROLE_ELEMENTALIST]
+        local allowUpgradedEffects = ROLE.ConvarTierUpgrades:GetBool()
+
+        if ROLE.ConvarPyroUpgrades:GetBool() then
```
Comment (NoxxFlame) [inline lua/customroles/elementalist/cl_elementalist.lua:119]: It might be worth putting a short description of each effect here for clarify

**Review by NoxxFlame — APPROVED** [review]
A couple of design comments but the code LGTM

File: materials/vgui/ttt/roles/elm/elementalist.png
Comment (minin43) [reply materials/vgui/ttt/roles/elm/elementalist.png:1]: I mean, yea they don't show up really well at all in-game, but they also don't really show up in game, so, no difference really

File: lua/customroles/elementalist/cl_elementalist.lua (hunk 16 above)
Comment (minin43) [reply lua/customroles/elementalist/cl_elementalist.lua:119]: The description might start getting raaaather lengthy with descriptions for each. The upgrades in the shop each have a description for the effect, having them here might be somewhat redundant. IDK.

**Pull request merged**